Avoid updating hidden page timer throttling interval for suspended processes

bnham opened the pull request:
#### 199ca3488158b61e977d8176fba7fcc8fd41a92e
<pre>
Avoid updating hidden page timer throttling interval for suspended processes
<a href="https://bugs.webkit.org/show_bug.cgi?id=270236">https://bugs.webkit.org/show_bug.cgi?id=270236</a>
<a href="https://rdar.apple.com/problem/123773191">rdar://problem/123773191</a>

Reviewed by Chris Dumez.

Currently, every time a tab is created or destroyed, we update the hidden page timer throttling
interval and broadcast the change to all processes, including suspended processes. While this
doesn&apos;t cause a suspended process to resume, it can cause messages to build up in IPC queues that
never get drained. This is because the receiver might be a suspended WebContent process.

To avoid this, skip sending the timer throttling interval to suspended processes and defer the
update until the resume IPC.

* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::hiddenPageThrottlingAutoIncreaseLimit):
(WebKit::WebProcessPool::updateHiddenPageThrottlingAutoIncreaseLimit):
* Source/WebKit/UIProcess/WebProcessPool.h:
(WebKit::WebProcessPool::sendToAllProcesses):
(WebKit::WebProcessPool::sendToAllProcessesForSession):
(WebKit::WebProcessPool::sendToAllRemoteWorkerProcesses):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::sendProcessDidResume):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setHiddenPageDOMTimerThrottlingIncreaseLimit):
(WebKit::WebProcess::processDidResume):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:

Canonical link: <a href="https://commits.webkit.org/275588@main">https://commits.webkit.org/275588@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1491f64b87dba9696b405c8d8f193ae57d9798b1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42226 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21244 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44620 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44821 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38342 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24444 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18577 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/34977 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42800 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18179 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/36364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/15913 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/15842 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46269 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/38420 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37733 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41636 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17042 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14026 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/40215 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18661 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9456 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18723 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18306 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/WebProcess.messages.in:
```diff
@@ -97,7 +97,7 @@ messages -> WebProcess LegacyReceiver NotRefCounted {
 
     PrepareToSuspend(bool isSuspensionImminent, MonotonicTime estimatedSuspendTime) -> ()
-    ProcessDidResume()
+    ProcessDidResume(Seconds hiddenPageThrottlingAutoIncreaseLimit)
```
Can't we leverage the existing SetHiddenPageDOMTimerThrottlingIncreaseLimit() instead of reusing `ProcessDidResume()`? I know this is one less IPC call upon resume but I'm not a fan of processDidResume being involved in timer throttling details.

szewai commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1757,8 +1757,12 @@ void WebProcessProxy::sendPrepareToSuspend(IsSuspensionImminent isSuspensionImmi
+    if (canSendMessage()) {
+        Seconds hiddenPageThrottlingAutoIncreaseLimit;
+        if (auto* pool = processPoolIfExists())
+            hiddenPageThrottlingAutoIncreaseLimit = pool->hiddenPageThrottlingAutoIncreaseLimit();
```
Do we want to unconditionally send the value whenever resume message is sent? I think resume message does not mean process is actually suspended.

bnham replied:
Hmm...unless we are spuriously resuming a lot, I don't think it should hurt to unconditionally send the new limit at process resume time. If it turns out to be an issue I can track whether we should send an update in an ivar.

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.h:
```diff
@@ -212,10 +213,10 @@ class WebProcessPool final
     void forEachProcessForSession(PAL::SessionID, const Function<void(WebProcessProxy&)>&);
-    template<typename T> void sendToAllProcesses(const T& message);
-    template<typename T> void sendToAllProcessesForSession(const T& message, PAL::SessionID);
+    template<typename T> void sendToAllProcesses(const T& message, ShouldSkipSuspendedProcesses = ShouldSkipSuspendedProcesses::No);
```
Do you mind if we don't add the default parameter? Are there just too many call sites to update?
Because frankly, it would be good to force people to think about whether or not the broadcast is useful for suspended processes and have them make a decision.

bnham replied:
Yeah, there are a lot of callsites. I'll file another bug to track investigate them. This particular one seemed to be triggered the most since it occurs on every tab open/close.

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.h:
```diff
@@ -212,10 +213,10 @@ class WebProcessPool final
-    template<typename T> void sendToAllProcesses(const T& message);
-    template<typename T> void sendToAllProcessesForSession(const T& message, PAL::SessionID);
+    template<typename T> void sendToAllProcesses(const T& message, ShouldSkipSuspendedProcesses = ShouldSkipSuspendedProcesses::No);
+    template<typename T> void sendToAllProcessesForSession(const T& message, PAL::SessionID, ShouldSkipSuspendedProcesses = ShouldSkipSuspendedProcesses::No);
```
ditto.

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.h:
```diff
@@ -212,10 +213,10 @@ class WebProcessPool final
+    template<typename T> void sendToAllProcessesForSession(const T& message, PAL::SessionID, ShouldSkipSuspendedProcesses = ShouldSkipSuspendedProcesses::No);
 
-    template<typename T> static void sendToAllRemoteWorkerProcesses(const T& message);
+    template<typename T> static void sendToAllRemoteWorkerProcesses(const T& message, ShouldSkipSuspendedProcesses = ShouldSkipSuspendedProcesses::No);
```
ditto.

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -1878,15 +1878,19 @@ void WebProcessPool::setFontAllowList(API::Array* array)
 }
 
-void WebProcessPool::updateHiddenPageThrottlingAutoIncreaseLimit()
+Seconds WebProcessPool::hiddenPageThrottlingAutoIncreaseLimit()
```
Should this be a const member function now?

bnham committed:
Avoid updating hidden page timer throttling interval for suspended processes
https://bugs.webkit.org/show_bug.cgi?id=270236
rdar://problem/123773191

Reviewed by Chris Dumez.

Currently, every time a tab is created or destroyed, we update the hidden page timer throttling
interval and broadcast the change to all processes, including suspended processes. While this
doesn't cause a suspended process to resume, it can cause messages to build up in IPC queues that
never get drained. This is because the receiver might be a suspended WebContent process.

To avoid this, skip sending the timer throttling interval to suspended processes and defer the
update until the resume IPC.

* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::hiddenPageThrottlingAutoIncreaseLimit):
(WebKit::WebProcessPool::updateHiddenPageThrottlingAutoIncreaseLimit):
* Source/WebKit/UIProcess/WebProcessPool.h:
(WebKit::WebProcessPool::sendToAllProcesses):
(WebKit::WebProcessPool::sendToAllProcessesForSession):
(WebKit::WebProcessPool::sendToAllRemoteWorkerProcesses):
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::sendProcessDidResume):
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::setHiddenPageDOMTimerThrottlingIncreaseLimit):
(WebKit::WebProcess::processDidResume):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:

Canonical link: https://commits.webkit.org/275588@main
